module/zfs/dsl_destroy.c

```diff
@@ -763,8 +763,14 @@ dsl_destroy_head_check_impl(dsl_dataset_t *ds, int expected_holds)
 	if (ds->ds_is_snapshot)
 		return (SET_ERROR(EINVAL));
 
-	if (zfs_refcount_count(&ds->ds_longholds) != expected_holds)
+	dsl_dir_t *dd = ds->ds_dir;
+	mutex_enter(&dd->dd_activity_lock);
+	if (zfs_refcount_count(&ds->ds_longholds) != expected_holds +
+	    dd->dd_activity_count) {
+		mutex_exit(&dd->dd_activity_lock);
 		return (SET_ERROR(EBUSY));
+	}
+	mutex_exit(&dd->dd_activity_lock);
```

Comment:
I think that if we change the semantic to ensure that we aren't waiting while the fs is not mounted, we can simplify some of this code, because we can't destroy it while it's mounted.